databroker/core.py

```diff
@@ -993,6 +1116,93 @@ def to_dask(self):
+        """
+        if isinstance(partition, (tuple, list)):
+            return super().read_partition(partition)
+
+        # Unpack partition
+        i = partition['index']
+
+        if isinstance(partition['fill'], str):
+            filler = self.fillers[partition['fill']]
+        else:
+            filler = partition['fill']
+
+        # Partition size is the number of pages in the partition.
+        if partition['partition_size'] == 'auto':
+            partition_size = 5
+        elif isinstance(partition['partition_size'], int):
+            partition_size = partition['partition_size']
+        else:
+            raise ValueError(f"Invalid partition_size {partition['partition_size']}")
+
+        if self._partitions is None:
+            self._load_partitions(partition_size)
+        try:
+            try:
+                return [filler(name, doc) for name, doc in self._partitions[i]]
+            except event_model.UnresolvableForeignKeyError as err:
+                # Slow path: This error should only happen if there is an old style
+                # resource document that doesn't have a run_start key.
+                self._partitions[i:i] = self._missing_datum(err.key, self.PARTITION_SIZE)
+                return [filler(name, doc) for name, doc in self._partitions[i]]
+        except IndexError as e:
+            return []
+
+    def _missing_datum(self, datum_id, partition_size):
+
+        # Get the resource from the datum_id.
+        if '/' in datum_id:
+            resource_uid, _ = datum_id.split('/', 1)
+        else:
+            resource_uid = self._lookup_resource_for_datum(datum_id)
+        resource = self._get_resource(uid=resource_uid)
+
+        # Use rechunk datum pages to make them into pages of size "partition_size"
+        # and yield one page per partition.  Rechunk might be slow.
+        datum_gen = self._get_datum_pages(resource['uid'])
+        partitions = [[('datum_page', datum_page)] for datum_page in
+                      event_model.rechunk_datum_pages(datum_gen, partition_size)]
+
+        # Check that the datum_id from the exception has been added.
+        def check():
+            for partition in partitions:
+                for name, datum_page in partition:
+                    if datum_id in datum_page['datum_id']:
+                        return True
+            return False
+
+        if not check():
+            raise
+
+        # Add the resource to the begining of the first partition.
+        partitions[0] = [('resource', resource)] + partitions[0]
+        self.npartitions += len(partitions)
+        return partitions
+
+    def _get_partition(self, partition):
+        return intake.container.base.get_partition(self.url, self.http_args,
+                                             self._source_id, self.container,
+                                             partition)
+
+class RemoteBlueskyEventStream(RemoteXarray):
+    def read_partition(self, partition):
```

Comment:
Can you add a comment on what this does? What would happen if `read_partition` weren't defined and we tried on the base classes instead?

Reply:
updated